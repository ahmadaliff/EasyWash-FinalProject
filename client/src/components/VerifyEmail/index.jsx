import PropTypes from 'prop-types';
import { useForm } from 'react-hook-form';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { FormattedMessage, injectIntl } from 'react-intl';

import InputRHF from '@components/InputRHF';

import { actionSetStep } from '@pages/Register/actions';
import { selectIsVerify } from '@containers/Client/selectors';
import { selectEmail, selectStep } from '@pages/Register/selectors';
import { actionHandleSendEmailVerify } from '@containers/Client/actions';

import classes from '@components/VerifyEmail/style.module.scss';

const VerifyEmail = ({ email, step, isVerified, intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const {
    handleSubmit,
    register,
    formState: { errors },
  } = useForm();

  const onSubmit = (data) => {
    if (!email || data.email !== email) dispatch(actionHandleSendEmailVerify(data));
    else if (isVerified) dispatch(actionSetStep(step + 2));
    else dispatch(actionSetStep(step + 1));
  };
  return (
    <form onSubmit={handleSubmit(onSubmit)} className={classes.form} data-testid="verify-email-form">
      <InputRHF
        input={{
          name: 'email',
          required: formatMessage({ id: 'app_user_email_require_message' }),
          type: 'text',
          label: formatMessage({ id: 'app_user_email' }),
          pattern: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+.[A-Z]{2,}$/i,
          messagePatern: formatMessage({ id: 'app_user_email_pattern_message' }),
          value: email,
        }}
        register={register}
        errors={errors}
      />
      <button type="submit" className={classes.buttonSubmit} data-testid="submit-button">
        <FormattedMessage id="app_next" />
      </button>
    </form>
  );
};

VerifyEmail.propTypes = {
  intl: PropTypes.object,
  step: PropTypes.number,
  email: PropTypes.string,
  isVerified: PropTypes.bool,
};

const mapStateToProps = createStructuredSelector({
  step: selectStep,
  email: selectEmail,
  isVerified: selectIsVerify,
});

export default injectIntl(connect(mapStateToProps)(VerifyEmail));
