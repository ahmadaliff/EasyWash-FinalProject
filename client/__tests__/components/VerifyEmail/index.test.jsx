import VerifyEmail from '@components/VerifyEmail';
import { render, fireEvent, screen } from '@utils/testHelper';
import { act } from 'react-dom/test-utils';

let wrapper;
let wrapperVerified;
let mockVerify = false;

const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));

jest.mock('@pages/Register/selectors', () => ({
  selectEmail: () => 'test@example.com',
  selectStep: () => 1,
}));
jest.mock('@containers/Client/selectors', () => ({
  selectIsVerify: () => mockVerify,
}));

const mockProps = {
  intl: { formatMessage: jest.fn() },
};

describe('VerifyEmail Component', () => {
  describe('with unverified email', () => {
    beforeEach(() => {
      act(() => {
        wrapper = render(<VerifyEmail {...mockProps} />);
      });
      jest.clearAllMocks();
    });
    test('Correct render', () => {
      const { getByTestId } = wrapper;
      expect(getByTestId('verify-email-form')).toBeInTheDocument();
      expect(screen.getByLabelText('Email')).toBeInTheDocument();
      expect(getByTestId('submit-button')).toBeInTheDocument();
    });

    test('Handles form submission when email is not verified', async () => {
      const { getByTestId } = wrapper;
      const inputEmail = screen.getByLabelText('Email');
      const submitButton = getByTestId('submit-button');
      fireEvent.change(inputEmail, { target: { value: 'new-email@example.com' } });

      await act(() => {
        fireEvent.click(submitButton);
      });
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
    test('Handles form submission when email is not verified and same as selector', async () => {
      const { getByTestId } = wrapper;
      const inputEmail = screen.getByLabelText('Email');
      const submitButton = getByTestId('submit-button');
      fireEvent.change(inputEmail, { target: { value: 'test@example.com' } });

      await act(() => {
        fireEvent.click(submitButton);
      });
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
  });
  describe('with verified email', () => {
    beforeEach(() => {
      mockVerify = true;
      act(() => {
        wrapperVerified = render(<VerifyEmail {...mockProps} />);
      });
      jest.clearAllMocks();
    });
    test('Handles form submission when email is verified', async () => {
      const { getByTestId } = wrapperVerified;
      const inputEmail = screen.getByLabelText('Email');
      const submitButton = getByTestId('submit-button');
      fireEvent.change(inputEmail, { target: { value: 'test@example.com' } });
      await act(() => {
        fireEvent.click(submitButton);
      });
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
    test('Matches snapshot', () => {
      expect(wrapperVerified).toMatchSnapshot();
    });
  });
});
