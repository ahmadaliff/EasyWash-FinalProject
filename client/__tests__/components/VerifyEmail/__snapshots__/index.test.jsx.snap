// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerifyEmail Component with verified email Matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="form"
        data-testid="verify-email-form"
      >
        <div
          class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
          data-testid="inputRHF"
        >
          <label
            class="inputLabel"
            for="email"
          >
            Email
            <input
              class="input undefined"
              id="email"
              name="email"
              type="text"
              value="test@example.com"
            />
          </label>
        </div>
        <button
          class="buttonSubmit"
          data-testid="submit-button"
          type="submit"
        >
          Lanjut
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="form"
      data-testid="verify-email-form"
    >
      <div
        class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
        data-testid="inputRHF"
      >
        <label
          class="inputLabel"
          for="email"
        >
          Email
          <input
            class="input undefined"
            id="email"
            name="email"
            type="text"
            value="test@example.com"
          />
        </label>
      </div>
      <button
        class="buttonSubmit"
        data-testid="submit-button"
        type="submit"
      >
        Lanjut
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
